### PR TITLE
Oxidizing and Waxing API

### PR DESCRIPTION
**Pull request by IThundxr**
Very simple, 2 non invasive patches & 2 extra methods to easily insert into the map

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/common/CommonHooks.java [hunk 1]
```diff
@@ -1360,4 +1362,30 @@ public static void onChunkUnload(PoiManager poiManager, ChunkAccess chunkAccess)
...
         }
     }
+
+    /**
+     * Registers a before and after blocks that can oxidize and de-oxidize
+     *
+     * @param before block with less oxidization
+     * @param after block with more oxidization
+     */
+    public static void registerOxidizableBlock(Block before, Block after) {
```
Comment (FiniteReality) [inline src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: I feel this would fit better in WeatheringCopper. Furthermore, as the HashBiMap is not thread-safe, we should document that this should be done on the main thread (e.g. using FMLCommonSetupEvent) to avoid concurrent modification exceptions.

File: patches/net/minecraft/world/item/HoneycombItem.java.patch [hunk 2]
```diff
@@ -0,0 +1,28 @@
+--- a/net/minecraft/world/item/HoneycombItem.java
++++ b/net/minecraft/world/item/HoneycombItem.java
+@@ -2,6 +_,7 @@
+ 
+ import com.google.common.base.Suppliers;
+ import com.google.common.collect.BiMap;
++import com.google.common.collect.HashBiMap;
```
Comment (FiniteReality) [inline patches/net/minecraft/world/item/HoneycombItem.java.patch]: You shouldn't modify imports in patches.

File: patches/net/minecraft/world/level/block/WeatheringCopper.java.patch [hunk 3]
```diff
@@ -0,0 +1,28 @@
+--- a/net/minecraft/world/level/block/WeatheringCopper.java
++++ b/net/minecraft/world/level/block/WeatheringCopper.java
+@@ -2,6 +_,7 @@
+ 
+ import com.google.common.base.Suppliers;
+ import com.google.common.collect.BiMap;
++import com.google.common.collect.HashBiMap;
```
Comment (FiniteReality) [inline patches/net/minecraft/world/level/block/WeatheringCopper.java.patch]: Again, you shouldn't modify imports in patches.

File: src/main/java/net/neoforged/neoforge/common/CommonHooks.java [hunk 4]
```diff
@@ -1360,4 +1362,30 @@ public static void onChunkUnload(PoiManager poiManager, ChunkAccess chunkAccess)
...
+        WeatheringCopper.NEXT_BY_BLOCK.get().put(before, after);
+    }
+
+    /**
+     * Registers a before and after blocks that can be waxed and unwaxed
+     *
+     * @param before the unwaxed block
+     * @param after the waxed block
+     */
+    public static void registerWaxableBlock(Block before, Block after) {
```
Comment (FiniteReality) [inline src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: Likewise, I feel this would fit better in its own type. Maybe we should create a `WaxableBlockRegistry` which contains this logic, and then change the patch to something like:
```diff
diff --git a/patches/net/minecraft/world/item/HoneycombItem.java.patch b/patches/net/minecraft/world/item/HoneycombItem.java.patch
new file mode 100644
index 0000000000..430a321653
--- /dev/null
+++ b/patches/net/minecraft/world/item/HoneycombItem.java.patch
@@ -0,0 +1,28 @@
+--- a/net/minecraft/world/item/HoneycombItem.java
++++ b/net/minecraft/world/item/HoneycombItem.java
+@@ -21,7 +_,7 @@
+ 
+ public class HoneycombItem extends Item implements SignApplicator {
+-    public static final Supplier<BiMap<Block, Block>> WAXABLES = Suppliers.memoize(
++    public static final Supplier<BiMap<Block, Block>> WAXABLES = net.neoforged.something.or.other.WaxableBlockRegistry::getWaxables;
++    public static final Supplier<BiMap<Block, Block>> VANILLA_WAXABLES = Suppliers.memoize(
+         () -> ImmutableBiMap.<Block, Block>builder()
+                 .put(Blocks.COPPER_BLOCK, Blocks.WAXED_COPPER_BLOCK)
+                 .put(Blocks.EXPOSED_COPPER, Blocks.WAXED_EXPOSED_COPPER)
+                 .put(Blocks.WEATHERED_COPPER, Blocks.WAXED_WEATHERED_COPPER)
+@@ -58,7 +_,7 @@
+                 .put(Blocks.EXPOSED_COPPER_BULB, Blocks.WAXED_EXPOSED_COPPER_BULB)
+                 .put(Blocks.WEATHERED_COPPER_BULB, Blocks.WAXED_WEATHERED_COPPER_BULB)
+                 .put(Blocks.OXIDIZED_COPPER_BULB, Blocks.WAXED_OXIDIZED_COPPER_BULB)
+                 .build()
+     );
+-    public static final Supplier<BiMap<Block, Block>> WAX_OFF_BY_BLOCK = Suppliers.memoize(() -> WAXABLES.get().inverse());
++    public static final Supplier<BiMap<Block, Block>> WAX_OFF_BY_BLOCK = net.neoforged.something.or.other.WaxableBlockRegistry::getReverseWaxables;
++    public static final Supplier<BiMap<Block, Block>> VANILLA_WAX_OFF_BY_BLOCK = Suppliers.memoize(() -> WAXABLES.get().inverse());
+ 
```

**Review by FiniteReality — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/common/CommonHooks.java (hunk 4 above)
Comment (FiniteReality) [reply src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: Alternatively, if we're moving to using the data map approach we could keep this class the same and simply deprecate the fields, pointing users towards the data map.

File: .editorconfig
Comment (ApexModder) [inline .editorconfig:1]: Why is there a .editorconfig file included with this PR?

Feel like this kind of thing should be in its own PR with more discussion on what formatting settings should be set.

Comment (IThundxr) [reply .editorconfig:1]: Intellij keeps wildcarding my imports and it's happened like thrice now, this is simply to prevent it from doing that (spotless already yells at you for it), so i didn't see any worth of making a whole new pr for 3 lines, anything else needed can be added later (to comply with spotless or whatever)

Comment (ApexModder) [reply .editorconfig:1]: Sounds like you have some auto import optimization setting enabled in IDEA, I would go ahead and disable that setting.

Could maybe be one of the following
_Settings > Editor > General > Auto Import_
- `Add unambiguous imports on the fly`
- `Optimize imports on the fly`

But yeah I think something like a `.editorconfig` which affects everyones idea settings should have more discussion/work around than being included in some random PR to fix a small annoyance.

Comment (embeddedt) [reply .editorconfig:1]: @IThundxr Pretty sure the setting to disable wildcard imports can be set per-project in your own settings

Comment (IThundxr) [reply .editorconfig:1]: > Sounds like you have some auto import optimization setting enabled in IDEA, I would go ahead and disable that setting.

no, the issue is 
image
https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcard-imports

Comment (IThundxr) [reply .editorconfig:1]: > @IThundxr Pretty sure the setting to disable wildcard imports can be set per-project in your own settings

yeah i know, just added it as a editorconfig as this seems like something everyone would want as otherwise spotless will complain, doesn't really make sense to tell everyone to disable it on their end when editorconfig can do it by default

Comment (ApexModder) [reply .editorconfig:1]: > just added it as a editorconfig as this seems like something everyone would want

While that maybe the case this is the wrong place to be putting it.
It should be in its own PR, this isn't a "formatter change" PR after all

Comment (IThundxr) [reply .editorconfig:1]: i guess so

File: src/main/java/net/neoforged/neoforge/common/CommonHooks.java (hunk 4 above)
Comment (IThundxr) [reply src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: yeah, i'll probably create datamaps for both these and deprecate the fields

**Comment by Spinoscythe** [comment]
I would personally say move this to an event or data map

File: src/main/java/net/neoforged/neoforge/common/CommonHooks.java (hunk 4 above)
Comment (lukebemish) [reply src/main/java/net/neoforged/neoforge/common/CommonHooks.java]: Those fields are all just `Supplier`s -- could you not just direct the supplier towards a map derived from the datamap?

**Comment by lukebemish** [comment]
I'd say this should definitely be a datamap if the only API being exposed is adding block pairs to those maps. Personally, I'd keep the suppliers around but have their contents redirect to some map derived from the datamap, but I can see the argument for just deprecating them and having people use the datamap manually. Redirecting the suppliers seems safer though.
